docs/INTERNALS.md

```diff
@@ -0,0 +1,319 @@
+
+## Terminology
+
+### KAPT
+
+KAPT is Kotlin’s annotation processing tool and allows using Java annotation processors for Kotlin.
+Anvil generates code that the Dagger 2 annotation processor must see and the integration with KAPT
+is crucial.
+
+### Hint
+
+Anvil generates “hints” for all contributed types. Hints make Anvil faster, because Anvil only
+needs to search in the hint package to find all contributions.
+
+### `CodeGenerator`
+
+A `CodeGenerator` is part of Anvil’s API and allows third parties to extend Anvil. Code generators
+run in the compiler frontend and generate Kotlin code.
+
+### `ClassReference`
+
+A `ClassReference` is an abstraction for the PSI and Descriptor APIs from the Kotlin compiler.
+During different stages of compilation the compiler uses one or the other API. `ClassReference`
+avoids different code branches and hides the necessary implementation details.
+
+### PSI
+
+Program Structure Interface (PSI) is an abstract syntax tree used in the frontend of the compiler
+to parse and represent source code.
+
+### Descriptors
+
+Descriptors are used in the old compiler backend (and partially frontend) to represent resolved
+types. After a type is resolved it doesn’t change anymore. A common use case is querying a type
+from a dependency that isn’t part of the compilation. Instead of having access to the source code
+(PSI) we rely on a descriptor. Descriptors themselves are an abstraction for Kotlin code,
+Java code and bytecode.
+
+## Gradle Plugin
+
+Like most compiler plugins Anvil comes with its own Gradle plugin handling all the ceremony to
+set up the compiler plugin. The Gradle plugin gives you options to configure Anvil through
+the `anvil { .. }` DSL, adds configurations to integrate custom code generators, e.g.
+
+```groovy
+dependencies {
+  anvil project(':sample:code-generator')
+  implementation project(':sample:annotation')
+}
+```
+
+adds the Anvil `annotations` artifact to the compile classpath and applies several fixes and
+workarounds to make the compiler plugin work, e.g. in incremental build scenarios:
+
+* [KT-38570](https://youtrack.jetbrains.com/issue/KT-38570)
+* [KT-38576](https://youtrack.jetbrains.com/issue/KT-38576)
+
+## KAPT Specifics
+
+Anvil makes dependency injection for Dagger 2 easier. Dagger 2 is an annotation processor and
+therefore Anvil must integrate with KAPT. The way KAPT works is that it adds two more Gradle tasks
+to a Gradle module before the Kotlin compilation task runs. The flow is similar to this:
+
+```
+generate Java stubs → run annotation processor → run Kotlin compilation → run Java compilation
+```
+
+It’s important to note that the stub generating task is a Kotlin compilation task and in this
+Gradle module Anvil runs twice: during stub generation and during Kotlin compilation. That’s
+crucial, because Anvil needs to generate code before Dagger runs in the annotation processing task.
+
+## Anvil’s Concept
```

Comment:
"Concept" doesn't quite feel right here.

Maybe "Design Goals"?

Reply:
> This section doesn't describe the design goals. Maybe a synonym like "approach", "process" or "technique" works better?

Aside from the first paragraph, this section feels focused on hints and performance, so perhaps it should be named based on that? e.g. "Hints", "Optimizing the Merge Phase", "Merging Contributions with Hints"

Reply:
I like "Merging Contributions with Hints" a lot.